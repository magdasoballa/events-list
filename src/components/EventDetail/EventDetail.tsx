import React, { useEffect, useState } from 'react';
import { useParams, useNavigate } from 'react-router-dom';
import { Card, CardContent, Typography, Button, Grid } from '@mui/material';
import axios from 'axios';
import { EventData } from '../../interfaces/eventInterfaces';
import { format } from 'date-fns';

const EventDetail: React.FC = () => {
    const { id } = useParams<{ id: string }>();
    const navigate = useNavigate();
    const [event, setEvent] = useState<EventData | null>(null);

    useEffect(() => {
        const fetchEvent = async (eventId: string) => {
            try {
                const response = await axios.get(`http://localhost:3001/events/${eventId}`);
                setEvent(response.data);
            } catch (error) {
                console.error('Error fetching event:', error);
            }
        };

        if (id) {
            fetchEvent(id);
        }
    }, [id]);

    const handleBackToList = () => {
        navigate('/events');
    };

    if (!event) {
        return <Typography variant="h6">Event not found!</Typography>;
    }

    const formatDate = (dateString: string) => {
        const date = new Date(dateString);
        if (isNaN(date.getTime())) {
            return 'Invalid Date';
        }
        return format(date, 'dd.MM.yyyy');
    };

    return (
        <Card>
            <CardContent style={{ padding: '16px' }}>
                <Grid container spacing={2}>
                    <Grid item xs={12} sm={6}>
                        {event?.title && (
                            <Typography variant="h5" style={{ fontSize: '1.2rem', fontWeight: 'bold', marginBottom: '8px' }}>
                                {event?.title}
                            </Typography>
                        )}
                        {event?.category && (
                            <Typography style={{ marginBottom: '8px' }}>
                                Category: {event?.category}
                            </Typography>
                        )}
                        {event?.date && (
                            <Typography style={{ marginBottom: '8px' }}>
                                Date: {formatDate(event?.date)}
                            </Typography>
                        )}
                        {event?.time && (
                            <Typography style={{ marginBottom: '8px' }}>
                                Time: {event?.time}
                            </Typography>
                        )}
                        {event?.description && (
                            <Typography style={{ marginBottom: '8px' }}>
                                Description: {event?.description}
                            </Typography>
                        )}
                        {event?.phoneNumber && (
                            <Typography style={{ marginBottom: '8px' }}>
                                Phone: {event?.phoneNumber}
                            </Typography>
                        )}
                        {event?.email && (
                            <Typography style={{ marginBottom: '8px' }}>
                                Email: {event?.email}
                            </Typography>
                        )}
                        {event?.location && (
                            <Typography style={{ marginBottom: '8px' }}>
                                Location: {event?.location}
                            </Typography>
                        )}
                        <Button
                            variant="contained"
                            color="primary"
                            style={{ marginTop: '16px' }}
                            onClick={handleBackToList}
                        >
                            Back to Events List
                        </Button>
                    </Grid>
                    <Grid item xs={12} sm={6}>
                        {event?.image && (
                            <img src={event.image} style={{ width: '100%', height: 'auto' }} alt="event" />
                        )}
                    </Grid>
                </Grid>
            </CardContent>
        </Card>
    );
};

export default EventDetail;
