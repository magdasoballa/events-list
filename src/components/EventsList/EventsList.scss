.list-wrapper {
    display: flex;
    align-items: center;
    justify-content: center;
    h4 {
        display: flex;
        align-items: center;
        justify-content: center;
        font-family: "Lato", Arial, sans-serif;
    }
}
