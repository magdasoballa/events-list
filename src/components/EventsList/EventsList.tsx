import React, { useEffect, useState } from 'react';
import { List, ListItem, ListItemText, ListItemAvatar, Avatar, Typography, Pagination, Box, Select, MenuItem, FormControl, InputLabel, SelectChangeEvent, Button, CircularProgress } from '@mui/material';
import EventIcon from '@mui/icons-material/Event';
import { useNavigate } from 'react-router-dom';
import axios from 'axios';
import './EventsList.scss'
import { useDispatch, useSelector } from 'react-redux';
import { RootState } from '../../redux/store';
import { setEvents } from '../../redux/events/eventSlice';

const EventsList: React.FC = () => {
    const [currentPage, setCurrentPage] = useState<number>(1);
    const [itemsPerPage, setItemsPerPage] = useState<number>(10);
    const [loading, setLoading] = useState<boolean>(true);
    const navigate = useNavigate();
    const dispatch = useDispatch();
    const events = useSelector((state: RootState) => state.events.events);

    useEffect(() => {
        const fetchEvents = async () => {
            try {
                const response = await axios.get('http://localhost:3001/events');
                if (response.status === 200) {
                    dispatch(setEvents(response.data));
                } else {
                    console.error('Network response was not ok');
                }
            } catch (error) {
                console.error('Error fetching events:', error);
            } finally {
                setLoading(false);
            }
        };

        fetchEvents();
    }, [dispatch]);

    const handlePageChange = (event: React.ChangeEvent<unknown>, page: number) => {
        setCurrentPage(page);
    };

    const handleItemsPerPageChange = (event: SelectChangeEvent<number>) => {
        setItemsPerPage(Number(event.target.value));
        setCurrentPage(1);
    };

    const getPaginatedEvents = () => {
        const startIndex = (currentPage - 1) * itemsPerPage;
        const endIndex = startIndex + itemsPerPage;
        return events.slice(startIndex, endIndex);
    };

    const handleItemClick = (id: number) => {
        navigate(`/events/${id}`);
    };

    const handleAddEvent = () => {
        navigate('/add-event');
    };

    return (
        <div className="list-wrapper">
            <Box sx={{ padding: 2 }}>
                <Typography variant="h4" gutterBottom>
                    Events List
                </Typography>

                <FormControl variant="outlined" sx={{ marginBottom: 2, display: 'flex' }}>
                    <InputLabel id="items-per-page-label">Items per page</InputLabel>
                    <Select
                        labelId="items-per-page-label"
                        value={itemsPerPage}
                        onChange={handleItemsPerPageChange}
                        label="Items per page"
                    >
                        <MenuItem value={5}>5</MenuItem>
                        <MenuItem value={10}>10</MenuItem>
                        <MenuItem value={20}>20</MenuItem>
                        <MenuItem value={50}>50</MenuItem>
                    </Select>
                </FormControl>

                {loading ? (
                    <Box sx={{ display: 'flex', justifyContent: 'center', alignItems: 'center', height: '50vh' }}>
                        <CircularProgress />
                    </Box>
                ) : (
                    <>
                        <List>
                            {getPaginatedEvents().map(event => (
                                <ListItem key={event.id} alignItems="flex-start" button onClick={() => handleItemClick(event.id as number)}>
                                    <ListItemAvatar>
                                        <Avatar>
                                            <EventIcon />
                                        </Avatar>
                                    </ListItemAvatar>
                                    <ListItemText
                                        primary={event.title}
                                        secondary={
                                            <>
                                                <Typography
                                                    sx={{ display: 'inline' }}
                                                    component="span"
                                                    variant="body2"
                                                    color="text.primary"
                                                >
                                                    {event.date}
                                                </Typography>
                                                {" â€” " + event.description}
                                            </>
                                        }
                                    />
                                </ListItem>
                            ))}
                        </List>
                        <Button sx={{ width: '100%' }} variant="contained" color="primary" onClick={handleAddEvent}>
                            Add Event
                        </Button>
                        <Box sx={{ display: 'flex', justifyContent: 'center', marginTop: 2 }}>
                            <Pagination
                                count={Math.ceil(events.length / itemsPerPage)}
                                page={currentPage}
                                onChange={handlePageChange}
                                color="primary"
                            />
                        </Box>
                    </>
                )}
            </Box>
        </div>
    );
};

export default EventsList;
