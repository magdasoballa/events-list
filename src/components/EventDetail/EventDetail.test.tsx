import React from 'react';
import { render, fireEvent, waitFor, act, screen } from '@testing-library/react';
import axios from 'axios';
import { MemoryRouter, Route, Routes } from 'react-router-dom';
import EventDetail from './EventDetail';
import { Provider } from 'react-redux';
import store from '../../redux/store';
import EventsList from '../EventsList/EventsList';

jest.mock('axios');
const mockAxios = axios as jest.Mocked<typeof axios>;

const mockEvent = {
    id: '1',
    title: 'Event 1',
    category: 'Kultura',
    date: '01.07.2024',
    time: '10:00',
    description: 'Description of Event 1',
    phoneNumber: '123-456-1',
    email: 'sample@example.com',
    location: 'Location 1',
    image: 'http://example.com/sample.jpg',
};

const mockNavigate = jest.fn();
jest.mock('react-router-dom', () => ({
    ...jest.requireActual('react-router-dom'),
    useNavigate: () => mockNavigate,
}));

describe('EventDetail component', () => {
    beforeEach(() => {
        mockAxios.get.mockResolvedValue({
            status: 200,
            data: mockEvent,
        });
    });

    afterEach(() => {
        jest.clearAllMocks();
    });

    test('should render event details correctly and navigate on button click', async () => {
        await act(async () => {
            render(
                <Provider store={store}>
                    <MemoryRouter initialEntries={['/events/1']}>
                        <Routes>
                            <Route path="/events/:id" element={<EventDetail />} />
                            <Route path="/events" element={<EventsList />} />
                        </Routes>
                    </MemoryRouter>
                </Provider>
            );
        });

        await waitFor(() => {
            expect(screen.getByText('Event 1')).toBeInTheDocument();
        });

        const backButton = screen.getByRole('button', { name: 'Back to Events List' });
        expect(backButton).toBeInTheDocument();

        await act(async () => {
            fireEvent.click(backButton);
        });

        expect(mockNavigate).toHaveBeenCalled();
        expect(mockNavigate).toHaveBeenCalledWith('/events');
    });
});
