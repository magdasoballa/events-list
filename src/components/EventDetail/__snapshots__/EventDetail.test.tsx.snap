// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EventDetail component should render event details correctly 1`] = `
<div>
  <h6
    class="MuiTypography-root MuiTypography-h6 css-2ulfj5-MuiTypography-root"
  >
    Event not found!
  </h6>
</div>
`;

exports[`EventDetail component should render event details correctly and navigate on button click 1`] = `
<div>
  <div
    class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-bhp9pd-MuiPaper-root-MuiCard-root"
  >
    <div
      class="MuiCardContent-root css-46bh2p-MuiCardContent-root"
      data-testid="event"
      style="padding: 16px;"
    >
      <div
        class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-2 css-mhc70k-MuiGrid-root"
      >
        <div
          class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-sm-6 css-1b3l6lk-MuiGrid-root"
        >
          <h5
            class="MuiTypography-root MuiTypography-h5 css-ag7rrr-MuiTypography-root"
            style="font-size: 1.2rem; font-weight: bold; margin-bottom: 8px;"
          >
            Event 1
          </h5>
          <p
            class="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
            style="margin-bottom: 8px;"
          >
            Category: 
            Kultura
          </p>
          <p
            class="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
            style="margin-bottom: 8px;"
          >
            Date: 
            07.01.2024
          </p>
          <p
            class="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
            style="margin-bottom: 8px;"
          >
            Time: 
            10:00
          </p>
          <p
            class="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
            style="margin-bottom: 8px;"
          >
            Description: 
            Description of Event 1
          </p>
          <p
            class="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
            style="margin-bottom: 8px;"
          >
            Phone: 
            123-456-1
          </p>
          <p
            class="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
            style="margin-bottom: 8px;"
          >
            Email: 
            sample@example.com
          </p>
          <p
            class="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
            style="margin-bottom: 8px;"
          >
            Location: 
            Location 1
          </p>
          <button
            class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary css-sghohy-MuiButtonBase-root-MuiButton-root"
            style="margin-top: 16px;"
            tabindex="0"
            type="button"
          >
            Back to Events List
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
        </div>
        <div
          class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-sm-6 css-1b3l6lk-MuiGrid-root"
        >
          <img
            alt="event"
            src="http://example.com/sample.jpg"
            style="width: 100%; height: auto;"
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;
